### ISSUE TITLE
Format selection should not prefer unreliable 'tbr' metadata on YouTube

### ISSUE THREAD

**Issue by glenn-slayden** [ytdl-org/youtube-dl#14143]
The **"tbr"** value in the YouTube JSON information, also reported in the `YouTubeDL -F` option, contains the boilerplate values **1155** (for format 135) or **2200** and **2310** (for format 136) when the bitrate is actually unknown. These special sentinal values are thus unrelated to the actual bitrate of the video and should not be interpreted as such, especially when comparing to other formats with valid bitrate values.

This pattern can be seen by scrolling in the following files. which aggregate f135/f136 reports for 60,134 YouTube items. The lines are sorted by reported bitrate, and you can note the very large gaps in the rightmost "filesize" column which are sections corresponding to the singular suspicious values mentioned above. The number of files in these gap areas is too large for the corresponding value to always be exactly "1155", "2200", or "2310" by coincidence.

http://www.blobule.com/webshare/ytf-135.txt
http://www.blobule.com/webshare/ytf-136.txt

Furthermore, many of the items advertising the aforementioned values have been spot-checked after download, and thier bitrates found to be unrelated to the values shown.

YouTubeDL currently does not detect and ignore these special case values. This often causes the `-f bestvideo` selection, which prioritizes higher bitrate, to erroneously choose the wrong format.

Please adjust the `bestvideo` **format selection** heuristic so that it ignores the bitrate when these special numeric values are seen, and defers instead to a secondary mechanism, such as pixel dimension, for the `bestvideo` determination.

At a minimum, `bestvideo` should corroborate that its final selection is credible vis-a-vis the other candidates. For example, consider YouTube video **Ot5jszj_ny8**, for which youtube-dl reports the following formats:

	format code  extension  resolution note
	139          m4a        audio only DASH audio   49k , m4a_dash container, mp4a.40.5@ 48k (22050Hz)
	140          m4a        audio only DASH audio  128k , m4a_dash container, mp4a.40.2@128k (44100Hz)
	160          mp4        256x144    DASH video  108k , avc1.4d400b, 30fps, video only
	134          mp4        640x360    DASH video  117k , avc1.4d401e, 30fps, video only
	133          mp4        426x240    DASH video  242k , avc1.4d400c, 30fps, video only
	136          mp4        1280x720   DASH video  365k , avc1.4d401f, 30fps, video only
	135          mp4        854x480    DASH video 1155k , avc1.4d4014, 30fps, video only
	17           3gp        176x144    small , mp4v.20.3, mp4a.40.2@ 24k
	36           3gp        320x180    small , mp4v.20.3, mp4a.40.2
	43           webm       640x360    medium , vp8.0, vorbis@128k
	18           mp4        640x360    medium , avc1.42001E, mp4a.40.2@ 96k
	22           mp4        1280x720   hd720 , avc1.64001F, mp4a.40.2@192k (best)

Here `-f bestvideo` will download format **135**, which, at 854×480, has an actual video bitrate of only **145 kb/s** (according to MediaInfo),  instead of format **136**, which is indeed 1280×720 with a considerably better rate of **169 kb/s**. Even with different compression levels, it is simply not credible that a format with **55% fewer pixels** could have a **bitrate 3x higher** for the exact same source content.

Another case, YouTube **MUMlwUe-BCo** was reported in #14010:

    136          mp4        1280x720   DASH video 1114k , avc1.4d401f, 30fps, video only
    135          mp4        854x480    DASH video 1155k , avc1.4d4014, 30fps, video only

Again here, format **136** was dis-preferred by `-f bestvideo`, but after downloading, the true bitrates were found to be:

- **124 kb/s** (as opposed to "1155") for `-f 135`
- **128 kb/s** (as opposed to "1114") for `-f 136`

**Comment by glenn-slayden** [comment]
Though perhaps I should have written it all here, I have outlined further thoughts on this in a comment to #6018.

However, my point is most forcefully relevant to *this* issue, so I'm providing a [direct link to that comment](https://github.com/rg3/youtube-dl/issues/6018#issuecomment-327654211) here.

**Comment by ddawson** [comment]
Responding to what you said in #14010:

> Being stuck with these bugs, is there any way to detect these situations where the metadata makes blatantly incredulous claims such as in this case, a format with **55% fewer pixels** supposedly having a **bitrate 3x higher**?

Well, what about using the resolution instead of the bitrate? Is it ever the case that a larger stream has a lower quality? Within the same format, at least? Doesn't seem likely to me.

>Again, is it possible that, for formats '''135''' and '''136''' (others?), any JSON tbr values that start with "11xx" should be ignored

Not sure that's a good idea. There are actual streams with such high bitrates, and much higher.

Actually, I went looking for high bitrate videos. I found IZJ48x7LsHg, which has a "~~25 Mbit~~" 12.5 Mbit stream (at 3840x2160 resolution), so yeah. Also, the rate reported for format 135 in this case is 1146k (actual avg. rate 552k). A bit different from the supposedly boilerplate value. I think YouTube's encoder is simply giving bad values, which could be for a variety of reasons. You'd think an encoder could give an accurate value for the video it *just encoded*, but apparently not. Or the value is pulled from something else.

**Comment by fireattack** [comment]
Agreed (as I mentioned before: https://github.com/rg3/youtube-dl/issues/6018#issuecomment-312967136) that at least for YouTube, we should use resolution as the _first_ indicator for quality. AFAIK, YouTube never upscale video, so no concern on that respect; also, the resolution information from API is also very accurate and more reliable (compared to bitrate or filesize).

**Comment by glenn-slayden** [comment]
@ddawson Yes, I'm aware that my issue above isn't clear on proposing a definitive solution. First of all, as I explained [here](https://github.com/rg3/youtube-dl/issues/6018#issuecomment-327654211) (including some 5th-grade math), bitrate should be summarily disqualified from the `-f bestvideo` determination at the outset.

But besides that, referring to those tables I linked ([135](http://www.blobule.com/webshare/ytf-135.txt), [136](http://www.blobule.com/webshare/ytf-136.txt)), I had started to suspect that ***any*** "bitrate" value where the **rightmost column is blank must be considered unreliable**. Taken together, this means we always ignore bitrate altogether, and it's the latter signal--a blank value for the file size--that would raise the alarm bells. 

So from there options start to emerge. Basically, always ignore bitrate, and use reported filesize (from the rightmost column) instead, for the following two reasons:

* due to the nature of this task, filesize [***can only be a same or better indicator than bitrate***](https://github.com/rg3/youtube-dl/issues/6018#issuecomment-327654211)
* never using bitrate entails that bitrate will not even be used as a "fallback" in cases where filesize is not reported, but *this is actually a huge big fix*, since current evidence suggests that ***it's precisely those cases*** where bitrate is wholly unreliable In other words (and, in hindsight, as expected) the only cases where bitrate can be trusted are those where filesize--which is a mathematically superior version of it--is reported as well anyway.

***--- or ---***

- Just always use total number of pixels transmitted in one second:  **width × height × fps**. 

I favor the latter, and it's actually quite hard to debunk. Note that this formula takes advantage of the like-duration property of the `f bestvideo` task to just multiply-in the **fps** (as opposed to **fps × duration**, since again, duration cancels across all of the comparison candidates). For comparison amongst alternative versions of the same source material, this formula nails down the notion of **best** pretty irrefutably, leaving only issues of decoder efficiency/loss, which are likely to be swamped by the contribution of these 3 factors.

I challenge anyone to present a `-F` output for YouTube where the maximal **width × height × fps** entry does not correlate with the notion of "best." Failing the appearance of such, I suggest that the onus is on any defenders of the current "bitrate" behavior to refute the points above.

And do note that this proposal (and points in the footnote as well) are **only proposed for the YouTube platform**, as a workaround to address serious and known metadata discrepencies (as discussed) that are specific to that site.

----
n.b. A [prior argument](https://github.com/rg3/youtube-dl/issues/11200#issuecomment-260555583) has suggested that the current behavior won't be changed because people will never "reach a consensus... [on] quality." To this, I respectfully point out that the ***luxury of consensus only applies when based on reliable choice***. In this case, the YouTube bug of failing to produce reliable bitrate information--nay worse, covertly providing it with falsified values--has a tangible, calamtous effect on the `-f bestvideo` result that, as a demonstrable error, falls clearly outside the regime of fair and legitimate opinion or "consensus."

Nobody's personal desire when using `bestvideo` is ever to receive a random file, which is part of the current behavior. The bitrates are not ***intended*** to be incorrect, and delivering an unintended result, by definition, is a bug. So it doesn't seem like using the 'variety of people's opinions' as an excuse for a mundane technological error with a deterministic cause is quite consistent this app's highest principles of excellence

**Comment by glenn-slayden** [comment]
In case anyone wants a temporary workaround for the **YouTube-specific bug** under discussion here, where **specifying** `-f bestvideo` **downloads a video format which is obviously not the best available quality**, I found that nulling out the `tbr` value from the format dictionary solves the problem for me:

**youtube_dl\extractor\common.py (1074)**

    for f in formats:
         # Automatically determine tbr when missing based on abr and vbr (improves
         # formats sorting in some cases)

    -    if 'tbr' not in f and f.get('abr') is not None and f.get('vbr') is not None:
    -        f['tbr'] = f['abr'] + f['vbr']

    +    #if 'tbr' not in f and f.get('abr') is not None and f.get('vbr') is not None:
    +    #    f['tbr'] = f['abr'] + f['vbr']
    +    f['tbr'] = None

**Before: (current behavior)**
Errors include **133** is better than **134** (it's not), and **135** is better than **136** (no chance, and in fact false).

    [info] Available formats for Ot5jszj_ny8:
    format code  extension  resolution note
    139          m4a        audio only DASH audio   49k , m4a_dash container, mp4a.40.5@ 48k (22050Hz)
    140          m4a        audio only DASH audio  128k , m4a_dash container, mp4a.40.2@128k (44100Hz)
    160          mp4        256x144    DASH video  108k , avc1.4d400b, 30fps, video only
    134          mp4        640x360    DASH video  117k , avc1.4d401e, 30fps, video only
    133          mp4        426x240    DASH video  242k , avc1.4d400c, 30fps, video only
    136          mp4        1280x720   DASH video  365k , avc1.4d401f, 30fps, video only
    135          mp4        854x480    DASH video 1155k , avc1.4d4014, 30fps, video only
    17           3gp        176x144    small , mp4v.20.3, mp4a.40.2@ 24k
    36           3gp        320x180    small , mp4v.20.3, mp4a.40.2
    43           webm       640x360    medium , vp8.0, vorbis@128k
    18           mp4        640x360    medium , avc1.42001E, mp4a.40.2@ 96k
    22           mp4        1280x720   hd720 , avc1.64001F, mp4a.40.2@192k (best)

**After: (with 3-line hack shown above)**

    [info] Available formats for Ot5jszj_ny8:
    format code  extension  resolution note
    139          m4a        audio only DASH audio , m4a_dash container, mp4a.40.5@ 48k (22050Hz)
    140          m4a        audio only DASH audio , m4a_dash container, mp4a.40.2@128k (44100Hz)
    160          mp4        256x144    DASH video , avc1.4d400b, 30fps, video only
    133          mp4        426x240    DASH video , avc1.4d400c, 30fps, video only
    134          mp4        640x360    DASH video , avc1.4d401e, 30fps, video only
    135          mp4        854x480    DASH video , avc1.4d4014, 30fps, video only
    136          mp4        1280x720   DASH video , avc1.4d401f, 30fps, video only
    17           3gp        176x144    small , mp4v.20.3, mp4a.40.2@ 24k
    36           3gp        320x180    small , mp4v.20.3, mp4a.40.2
    43           webm       640x360    medium , vp8.0, vorbis@128k
    18           mp4        640x360    medium , avc1.42001E, mp4a.40.2@ 96k
    22           mp4        1280x720   hd720 , avc1.64001F, mp4a.40.2@192k (best)

Notice that now, not only is the worst-to-best ordering is more in line with expectations, but those grossly unreliable bitrate values are also totally gone from the table listing.

**Comment by Hrxn** [comment]
Agreed.

> width × height × fps

Yes, I concur. And true to the old adage _"more pixels, more better"_..

But one issue remains here:

MP4/AVC vs. WEBM/VP9x

**Comment by rautamiekka** [comment]
^
>MP4/AVC vs. WEBM/VP9x

WEBM tends to have worse quality than MP4, but it varies: sometimes WEBM's quality at same or even higher resolution is much worse than MP4 and the file size could be anything, other times not much worse but the file size could be anything; I've had many WEBM files where WEBM was superior in the sense the quality wasn't that much lower it would make a real difference, but the file size was smaller, but also the same happened where the file is much bigger.

**Comment by glenn-slayden** [comment]
> I think YouTube's encoder is simply giving bad values, which could be for a variety of reasons. You'd think an encoder could give an accurate value for the video it just encoded, but apparently not. 

I think it's more likely that the reported `tbr` value is normally comptued from the data size field which appears in the rightmost column. When that value is not available for some reason, the `tbr` computation falls back to some kind of extremely crude approximation, possibly using values from a lookup table. This would explain the clustering of `tbr` values that I noted, and would also suggest that ***any*** `tbr` value where the filesize is blank should be ignored.

**Comment by ddawson** [comment]
> When the data size is not available for some reason, the `tbr` computation falls back to some kind of extremely crude approximation, possibly using values from a lookup table.

You might be right about that. But how do you explain this one (from aforementioned IZJ48x7LsHg):

```
135          mp4        854x480    DASH video 1146k , avc1.4d401e, 25fps, video only, 19.26MiB
```

Has a stated (and accurate) filesize, but given rate is *still* off by over 2x (it's actually ~552k). Ridiculous.

> ***whenever filesize is not reported, the 'tbr' value should be ignored***

That's okay, I suppose. Or rather, per my example, maybe it should just be completely ignored?

**Comment by glenn-slayden** [comment]
> maybe [YouTube-reported 'tbr' values] should just be completely ignored?

I agree 100%.

**Comment by lilydjwg** [comment]
As you can see in #25925, even when the actual tbr is higher, it doesn't mean the video quality is better: the codec may be more efficient.

From video 'https://www.youtube.com/watch?v=iygjJ8M7jnM', format 22 tbr reportedly at 362k and 367 kb/s by ffprobe, is better than format 18 tbr reportedly at 373k and 377 kb/s by ffprobe. The former is much clearer. (just view it fullscreen!)

I suggest sorting prefer dimensions than tbr / filesize, because video sites provide videos of better quality with higher (or equal) dimensions.

**Comment by glenn-slayden** [comment]
> @lilydjwg  wrote:<br>I suggest sorting prefer dimensions than tbr / filesize, because video sites provide videos of better quality with higher (or equal) dimensions.

Unfortunately, this might not be a solution either. Videos with larger pixel dimensions can have much lower resolution, typically by showing huge compression-artifact blocks... (note that I'm just stating a fact here, I don't have an actual case from yt).

**Comment by fireattack** [comment]
Videos with larger pixel dimensions *by definition* have *higher* resolution, I have no idea what you're saying.

Can these higher resolution versions being upscaled? Of course. But they're still "higher resolution" (and I don't think YT do up-scaling.)

**Comment by glenn-slayden** [comment]
@fireattack  Thanks, I clarified my post.

And a follow-up question: Is it possible on YT for the uploading user to **explicitly supply** multiple source video files (corresponding to different YT formats), or does YT always generate those various available formats itself?

If so, the uploading user might have provided up-scaled videos. But if it's the latter case, and that automatic process never up-scales, then it's true my original comment would not apply.

**Comment by ddawson** [comment]
> And a follow-up question: Is it possible on YT for the uploading user to **explicitly supply** multiple source video files (corresponding to different YT formats), or does YT always generate those various available formats itself?

Uploading multiple sources is not something YouTube provides for. If it were, then people could just upload completely different videos for different resolutions, which would be a mess for a variety of reasons.

**Comment by glenn-slayden** [comment]
@ddawson  thanks for that useful info.

**Comment by danny-wu** [comment]
Would it be possible to officially build in support for ignoring tbr, perhaps as a format option, or even just a site-specific CLI option like `--youtube-ignore-tbr`? There is precedent for this with the `--youtube-skip-dash-manifest` option.

I would like to ignore `tbr` for my youtube-dl archival systems, however I do not want to patch youtube-dl and have to keep updating the changes.

**Comment by dirkf** [comment]
Continued in [#33054](https://github.com/ytdl-org/youtube-dl/issues/33054#issuecomment-2649497605).
